spec/unit/provider/package_targetable_spec.rb

```diff
@@ -0,0 +1,54 @@
+require 'spec_helper'
+require 'puppet'
+require 'puppet/provider/package_targetable'
+require 'puppet/provider/package/gem'
+
+describe Puppet::Provider::Package::Targetable do
+  let(:provider) { Puppet::Type.type(:package).provide(:gem, :parent => described_class) }
+  let(:command)  { '/opt/bin/gem' }
+
+  describe "when prefetching" do
+    context "with a package without a command attribute" do
+      let(:resource) { Puppet::Type.type(:package).new(:name => 'noo', :provider => 'gem', :ensure => :present) }
+      let(:catalog)  { Puppet::Resource::Catalog.new }
+      let(:instance) { OpenStruct.new(:name => 'noo') }
```

Comment:
Could you construct a real `Puppet::Type.type(:package)` for a different provider?

Reply:
This tests whether `self.prefetch(packages)` in `Puppet::Provider::Package::Targetable` finds and passes `command` attributes in the catalog to the `self.instances` method of the associated provider. So what is unique here is whether `self.instances` receives `nil` or `/opt/bin/gem` and what `self.instances` returns is already covered by the spec tests in the provider(s).

I would like to add a context to test the combination of two (targetable) package resources, one with and one without a `command` to test that `self.instances` is called  instances is called both with `nil` and `/opt/bin/gem` but have not been able to construct a proper `let(:packages)` to pass to `provider.prefetch(packages)`

Reply:
Without using `to_s` in the provider, the provider (appears to) compare the `object_id` of each class ... which fails only when testing via rspec:

```
70297784197300 vs 70297513115040
```

Using `to_s ` in the provider, it compares the names of each class:

```
Puppet::Type::Package::ProviderGem vs Puppet::Type::Package::ProviderGem 
```

Reply:
Charlie suggest the need for `to_s` may caused by this executing multiple times: 

https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/type.rb#L1866